cmake/CmakeFunctionHelper.cmake

```diff
@@ -26,18 +26,27 @@
 set(executor_list "PolicyHandler<codeplay_policy>")
 #represent the list of supported index/increment type
 set(index_list "int" )
-#represent the list of supported data type.
+
+# BLAS_DATA_TYPES represents the list of supported data type.
 #Each data type in a data list determines the container types.
 #The container type for SYCLbackend is BufferIterator<${data}, codeplay_policy>
-set(data_list "float")
- #if double supported we add double as a data type
-if(DOUBLE_SUPPORT)
-  set(data_list "float" "double")
-endif()
 
 ## represent the list of bolean options
 set(boolean_list "true" "false")
 
+# Maps a user provided data type name to the C++ type
+# See BLAS_DATA_TYPES
+function(to_cpp_type output data)
```

Comment:
I don't really understand why this is a function, or what it does. Can't you just inline the `set` call where you use it? Then you don't need the `PARENT_SCOPE`

Reply:
In this PR the function doesn't really do anything, but it's important for new data types. For exaple, the user might pass in `_DBLAS_DATA_TYPES="float;half"`, but the actual C++ type would be `cl::sycl::half`.

Reply:
Oh, I see. Maybe it would be easier to have the old style flags (enable double, enable half etc.) and use the presence of those flags to construct the list internally, such that it happens once and you know its contents are correct?

Reply:
I don't know, I prefer to have a list of supported types instead of potentially many different flags. If you misspell the flag CMake will just warn of unused variables, whereas with wrong types in a list you'll get configuration or compilation errors. I think it's also easier to search for `BLAS_DATA_TYPE` (without the S) in the repo which will give you all the places where any type related shenaningans are happening, which really shouldn't be that many places.

Reply:
I don't really agree, but I don't have strong feelings. We could still move the processing of the list such that each element in it is guaranteed to be the correct SYCL type by this point in the code, then you don't need to jump out to this function each time. In theory, can I have `DATA_TYPES=half` without `float` and it will work?

Reply:
In theory, skipping `float` could work, but I didn't invest the necessary effort into it because I'm focusing on `cl::sycl::half`. I think it would probably be relatively easy to do, but there might be some hidden issues.

I removed the function for now and instead I just sanitize the filename: https://github.com/codeplaysoftware/sycl-blas/pull/245/commits/f83012489997567492d2508b6a3f0af99597604a . Note that this functionality has been there before, I've just made it into a function and extended the filtering to also cover `:`. `MAKE_C_IDENTIFIER` isn't suitable anymore because it also converts the `.` of the file extension.